runtime/include/http_session.h

```diff
@@ -0,0 +1,364 @@
+	http_session_close(session);
+	http_session_free(session);
+
+	return rc;
+}
+
+static inline int
+http_session_send_response(struct http_session *session, const char *response_content_type, void_cb on_eagain)
+{
+	assert(session != NULL);
+	assert(session->response_buffer.buffer != NULL);
+	assert(response_content_type != NULL);
+
+	int rc = 0;
+
+	struct vec_u8 *response_buffer = &session->response_buffer;
+
+	/* Send HTTP Response Header and Body */
+	rc = http_header_200_write(session->socket, response_content_type, response_buffer->length);
+	if (rc < 0) goto err;
+
+	rc = tcp_session_send(session->socket, (const char *)response_buffer->buffer, response_buffer->length,
+	                      on_eagain);
+	if (rc < 0) goto err;
+
+	http_total_increment_2xx();
+	rc = 0;
+
+done:
+	return rc;
+err:
+	debuglog("Error sending to client: %s", strerror(errno));
+	rc = -1;
+	goto done;
+}
+
+
+typedef void (*http_session_receive_request_egain_cb)(struct http_session *);
+
+/**
+ * Receive and Parse the Request for the current sandbox
+ * @return 0 if message parsing complete, -1 on error, -2 if buffers run out of space
+ */
+static inline int
+http_session_receive_request(struct http_session *session, http_session_receive_request_egain_cb on_eagain)
+{
+	assert(session != NULL);
+
+	int rc = 0;
+
+	struct vec_u8 *request_buffer = &session->request_buffer;
+	assert(request_buffer->capacity > 0);
+	assert(request_buffer->length < request_buffer->capacity);
+
+	while (!session->http_request.message_end) {
+		/* Read from the Socket */
+
+		/* Structured to closely follow usage example at https://github.com/nodejs/http-parser */
+		http_parser                *parser   = &session->http_parser;
+		const http_parser_settings *settings = http_parser_settings_get();
+
+		/* If header parsing is complete, resize using content-length */
+		if (session->http_request.header_end && session->http_request.body != NULL) {
+			int header_size = (uint8_t *)session->http_request.body - session->request_buffer.buffer;
+			assert(header_size > 0);
+			int required_size = header_size + session->http_request.body_length;
+
+			assert(required_size > 0);
+
+			if (required_size > request_buffer->capacity) {
+				uint8_t *old_buffer = request_buffer->buffer;
+				if (vec_u8_resize(request_buffer, required_size) != 0) {
+					debuglog("Failed to resize request vector to %d bytes\n", required_size);
+					goto err_nobufs;
+				}
+
+				if (old_buffer != request_buffer->buffer) {
+					/* buffer moved, so invalidate to reparse */
+					memset(&session->http_request, 0, sizeof(struct http_request));
+					http_parser_init(&session->http_parser, HTTP_REQUEST);
+					/* Set the session as the data the http-parser has access to */
+					session->http_parser.data = &session->http_request;
+				}
+			}
+		} else if (request_buffer->length == request_buffer->capacity) {
+			/* Otherwise, we have a huge header and should just grow */
```

Comment:
It's like a vector. The length member is the size that is used. The capacity member is the total size. If they're equal, the buffer is full. We don't know the amount of data the socket contains in this circumstance.

Reply:
Yes, is it guaranteed that `len` will increase monotonically? Meaning, can we not have the following kind of scenario?
len is 10; cap is 14;
len += 6; so now len is 16.

So, len never got a chance to be 14...

Reply:
It shouldn't be possible because we can only receive up to the amount of remaining buffer:
```
ssize_t bytes_received = recv(session->socket, &request_buffer->buffer[request_buffer->length], request_buffer->capacity - request_buffer->length, 0);
request_buffer->length += bytes_received;
```

I'll harden with some assertions though!